### PR TITLE
Implement ability to validate dashboard and pipeline specifications

### PR DESCRIPTION
**Pull request by philippjfr**
This PR now implements roughly the proposal I outlined https://github.com/holoviz/lumen/issues/301

The main goals were all achieved:

- [x] Add a lumen validate command that can provide useful validation without explicitly instantiating a full dashboard
- [x] Implement validate methods for all components
- [x] Validation errors are highly precise and informative including the ability to highlight the exact part of the specification where the error occurred.

The validate method has the following signature:

```python

    @classmethod
    def validate(cls, spec, context=None):
        """
        Validates the component specification given the validation context.

        Arguments
        -----------
        spec: dict
          The specification for the component being validated.
        context: dict
          Validation context contains the specification of all previously validated components,
          e.g. to allow resolving of references.

        Returns
        --------
        Validated specification.
        """
```

This allows the validation to a) validate the current spec and b) resolve any references. If an issue is found a `ValidationError` is raised. A ValidationError consists of three things:

1. A message describing what went wrong
2. If the field is invalid a suggestion like "Did you mean 'sources'?"
3. A printout of the part of the spec that is invalid and if possible a bold highlight of the specific key or value that caused the error. 

Most of the validation machinery is now handled by the `Component` and `MultiTypeComponent` base classes. In particular validation goes through a number of checks:

1. If the component is a `MultiTypeComponent` (i.e. a `Filter`, `Source`, `Transform`, `View` or `Variable`),  a `type` must be provided and the type is resolved to see if it is valid
4. Check if `_allowed_keys` is declared and if so validate all existing entries in the spec against the allowed keys
5. Check if `_required_keys` is declared and ensure that all required key are provided
6. Now iterate over all provided keys and do one of the following:
    a. Check if value is a reference and `_allow_refs=False`
    b. Check if key references a parameter that expects another `Component`, if so defer to its `validate` method
    c. Check if key references a parameter that expects a list of another `Component`, if so iterate over list and defer to them `validate` method for each item
    d. Check if value is a reference and ensure the reference can be resolved
    e. If `_validate_params=True` and the field maps to a parameter use the `Parameter._validate` method

Supersedes https://github.com/holoviz/lumen/pull/300

## To Do

- [ ] Allow `_validate_params` to accept a list of parameters to validate using regular parameter validation
- [ ] Go through all components and ensure that `_valid_keys`, `_required_keys`, `_allow_refs` and `_validate_params` are set correctly
- [ ] Review all `ValidationError` messages for readability
- [ ] Revert `ConfigDict`
- [ ] Write validation tests for all Component types
    - [ ] `lumen.dashboard.Auth`
    - [x] `lumen.dashboard.Config`
    - [x] `lumen.dashboard.Defaults`
    - [ ] `lumen.dashboard.Dashboard`
    - [ ] `lumen.filters.base.Filter`
    - [ ] `lumen.pipeline.Pipeline`
    - [ ] `lumen.sources.base.Source`
    - [x] `lumen.target.Download`
    - [x] `lumen.target.Facet`
    - [x] `lumen.target.Target`
    - [ ] `lumen.transforms.base.Transform`
    - [ ] `lumen.views.base.Download`
    - [ ] `lumen.views.base.View`
- [ ] Document validation

### CODE REVIEW

File: lumen/tests/validation/test_validation_error.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+from lumen.validation import ValidationError
+
+
+def test_only_bold_whole_words():
+    spec = {'title': 'Table', 'source': 'penguin', 'views': [{'type': 'table', 'table': 'penguins'}]}
+    msg = r'test\n\n    title: Table\n    source: \x1b[1mpenguin\x1b[0m\n    views:\n    - type: table\n      table: penguins\n'
+    try:
+        ValidationError("test", spec, "penguin")
```
Comment (philippjfr) [inline lumen/tests/validation/test_validation_error.py:8]:
```suggestion
        raise ValidationError("test", spec, "penguin")
```

Comment (philippjfr) [reply lumen/tests/validation/test_validation_error.py:8]: Right now, the assert never triggers.

Comment (Hoxbro) [reply lumen/tests/validation/test_validation_error.py]: Doh. Thank you. Fixed in  5154087.

**Comment by philippjfr** [comment]
Will merge at this point and carry on in smaller PRs.

**Pull request merged**